### PR TITLE
feat: implement TryFrom<PublicOrSecret> for Signed{Public,Secret}Key

### PR DESCRIPTION
**Pull request by link2xt**
This is less dangerous than checking
`is_public()` or `is_secret()`
and then using `into_secret()` and `into_public()` that may panic.

### CODE REVIEW

File: src/composed/signed_key/shared.rs [hunk 1]
```diff
@@ -246,6 +247,41 @@ impl PublicOrSecret {
...
+impl fmt::Display for TryFromPublicOrSecretError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Attempt to convert PublicOrSecret key to the wrong type")
+    }
+}
+
+impl std::error::Error for TryFromPublicOrSecretError {}
+
+impl TryFrom<PublicOrSecret> for SignedPublicKey {
+    type Error = TryFromPublicOrSecretError;
```
Comment (hko-s) [inline src/composed/signed_key/shared.rs:264]: Wouldn't it be easier and more ergonomic to use `pgp::errors::Error` here?

Comment (hko-s) [reply src/composed/signed_key/shared.rs:264]: Hmm. I never thought about how big `pgp::errors::Error` is, in terms of memory. I would have hoped it's minor.

Is there reason to consider this operation performance critical? The parsing that precedes this call should be more expensive by orders of magnitude.

I'd worry that a separate error type makes error handling more confusing for users of rPGP.

So my preference would be to reuse the regular rpgp error type, but this preference is not very strongly held.

**Comment by hko-s** [comment]
Putting aside the error-type question in my comment, I think this PR is good.

It would be nice to deprecate/remove the panic-y functions at some point.
Maybe deprecating them already now would be good?

**Comment by hko-s** [comment]
> It would be nice to deprecate/remove the panic-y functions at some point. Maybe deprecating them already now would be good?

Oh no. The deprecation is making CI very angry :scream_cat:

File: src/composed/signed_key/shared.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/composed/signed_key/shared.rs:266]: we probably need to split that error up into multiple suberrors anyway at some point, so this fine

File: src/composed/signed_key/shared.rs [hunk 2]
```diff
@@ -246,6 +249,41 @@ impl PublicOrSecret {
     }
 }
 
+/// Error returned when trying to convert `PublicOrSecret` key
+/// into the wrong type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct TryFromPublicOrSecretError;
```
Comment (dignifiedquire) [inline src/composed/signed_key/shared.rs:255]: we already have `thiserror` in the dependency tree, so please use that

**Review by dignifiedquire — APPROVED** [review]

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**